server/src/main/java/org/elasticsearch/client/transport/TransportClient.java

```diff
@@ -172,6 +173,7 @@ private static ClientTemplate buildTemplate(
             entries.addAll(searchModule.getNamedWriteables());
             entries.addAll(indicesModule.getNamedWriteables());
             entries.addAll(ClusterModule.getNamedWriteables());
+            entries.addAll(SystemIndexMigrationExecutor.getNamedWriteables());
```

Comment:
this fixes the test problem where transport client used in testing is missing SystemIndexMigrationTaskParams/State entries
I think these entries should be the same as in Node class
https://github.com/elastic/elasticsearch/blob/7.17/server/src/main/java/org/elasticsearch/node/Node.java#L538

How about the NamedXContentRegistry in TransportClient vs Node? The failure https://gradle-enterprise.elastic.co/s/4fqbspomp46qs/tests/:modules:reindex:internalClusterTest/org.elasticsearch.migration.SystemIndexMigrationIT/testSystemIndexMigrationCanBeInterruptedWithShutdown?top-execution=1 
was occuring when a test was trying to create an index with the client from ESIntegTest
https://github.com/elastic/elasticsearch/pull/84258/files#diff-ac55b6e8c817aafb7844d2584fe44afd82a8b06f8a248b94af642e5429fe577cR75

Reply:
but i am not sure this could work for other tests (featuremigrationIT for instance..)

Reply:
I confirmed this with @droberts195  it looks like the logic depends on getPerTestTransportClientRatio and might happen that the tests will pass if a node client was used instead of transport client

Reply:
> The failure https://gradle-enterprise.elastic.co/s/4fqbspomp46qs/tests/:modules:reindex:internalClusterTest/org.elasticsearch.migration.SystemIndexMigrationIT/testSystemIndexMigrationCanBeInterruptedWithShutdown?top-execution=1
was occuring when a test was trying to create an index with the client from ESIntegTest
https://github.com/elastic/elasticsearch/pull/84258/files#diff-ac55b6e8c817aafb7844d2584fe44afd82a8b06f8a248b94af642e5429fe577cR75

The above was incorrect. In fact, the failure was occuring when a feature upgrade is submitted. I managed to reproduce this in FeatureMigrationIT with the same logic we use in SystemIndexMigrationIT. 
1. add a listener that waits for the task to be created and blocks
2. submit a feature upgrade
3. wait for the task to be created until calling the get api
4. the failure will be visible in console
```
TransportSerializationException[Failed to deserialize response from handler [org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler/org.elasticsearch.action.ActionListenerResponseHandler@6be3d9d/org.elasticsearch.client.transport.TransportClientNodesService$RetryListener@2cd1aed]
]; nested: IllegalArgumentException[Unknown NamedWriteable category [org.elasticsearch.persistent.PersistentTaskParams]];
```